src/partition/cuda/partition_op.cu

```diff
@@ -0,0 +1,185 @@
+ * \file ndarray_partition.h
+ * \brief Operations on partition implemented in CUDA. 
+ */
+
+#include "../partition_op.h"
+
+#include <dgl/runtime/device_api.h>
+
+#include "../../array/cuda/dgl_cub.cuh"
+#include "../../runtime/cuda/cuda_common.h"
+
+using namespace dgl::runtime;
+
+namespace dgl {
+namespace partition {
+namespace impl {
+
+template<typename IdType> __global__ void _MapProcByRemainder(
+    const IdType * const index,
+    const int64_t num_index,
+    const int64_t num_proc,
+    IdType * const proc_id) {
+  const int64_t idx = blockDim.x*static_cast<int64_t>(blockIdx.x)+threadIdx.x;
+
+  if (idx < num_index) {
+    proc_id[idx] = index[idx] % num_proc;
+  }
+}
+
+template<typename IdType>
+__global__ void _MapProcByMaskRemainder(
+    const IdType * const index,
+    const int64_t num_index,
+    const IdType mask,
+    IdType * const proc_id) {
+  const int64_t idx = blockDim.x*static_cast<int64_t>(blockIdx.x)+threadIdx.x;
+
+  if (idx < num_index) {
+    proc_id[idx] = index[idx] & mask;
+  }
+}
+
+
+
+template <DLDeviceType XPU, typename IdType>
+std::pair<IdArray, NDArray>
+GeneratePermutationFromRemainder(
+        int64_t array_size,
+        int num_parts,
+        IdArray in_idx) {
+  std::pair<IdArray, NDArray> result;
+
+  const auto& ctx = in_idx->ctx;
+  auto device = DeviceAPI::Get(ctx);
+
+  const int64_t num_in = in_idx->shape[0];
+
+  CHECK_GE(num_parts, 1);
+  if (num_parts == 1) {
+    // no permutation
+    result.first = aten::Range(0, num_in, sizeof(IdType)*8, ctx);
+    result.second = aten::Full(num_in, num_parts, sizeof(int64_t)*8, ctx);
+
+    return result;
+  }
+
+  result.second = aten::Full(0, num_parts, sizeof(int64_t)*8, ctx);
+  int64_t * out_counts = static_cast<int64_t*>(result.second->data);
+  if (num_in == 0) {
+    // now that we've zero'd out_counts, nothing left to do for an empty
+    // mapping
+    return result;
+  }
+
+  const int64_t part_bits =
+      static_cast<int64_t>(std::ceil(std::log2(num_parts)));
+
+  // First, generate a mapping of indexes to processors
+  IdType * proc_id_in = static_cast<IdType*>(
+      device->AllocWorkspace(ctx, sizeof(IdType)*num_in));
+  {
+    const dim3 block(256);
+    const dim3 grid((num_in+block.x-1)/block.x);
+
+    if (num_parts < (1 << part_bits)) {
+      // num_parts is not a power of 2
+      _MapProcByRemainder<<<grid, block>>>(
+          static_cast<const IdType*>(in_idx->data),
+          num_in,
+          num_parts,
+          proc_id_in);
+      CUDA_CALL(cudaGetLastError());
+    } else {
+      // num_parts is a power of 2
+      _MapProcByMaskRemainder<<<grid, block>>>(
+          static_cast<const IdType*>(in_idx->data),
+          num_in,
+          static_cast<IdType>(num_parts-1),  // bit mask
+          proc_id_in);
+      CUDA_CALL(cudaGetLastError());
+    }
+  }
+
+  // then create a permutation array that groups processors together by
+  // performing a radix sort
+  IdType * proc_id_out = static_cast<IdType*>(
```

Comment:
Where to free this object?

Reply:
I switched all workspace allocations to use RAII via https://github.com/dmlc/dgl/pull/2825/files#diff-128b2340cfe395c91243eab36571e6f38af6b4eb0d0c3e01eaaba552995d8828 in 
69daf52, to prevent leaking workspace.